### ISSUE TITLE
Argo Exec needs to retry for non-transient error as well

### ISSUE THREAD

**Issue by Sharathmk99** [argoproj/argo-workflows#11749]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

Hi Team,

We are using Minio as our artifact storage, sometime due to network problem we get `dial tcp: i/o timeout`. This error is configured as `non-transient` error. Due to which retry is skipped.

Source code ref:
https://github.com/argoproj/argo-workflows/blob/master/workflow/artifacts/s3/s3.go#L81-L89

Error I'm getting when network is not stable,

```
2023-09-05T09:30:29.657599935Z stderr F time="2023-09-05T09:30:29.657Z" level=warning msg="Non-transient error: Get \"http://minio-service.kubeflow.svc.cluster.local:9000/mlpipeline/?location=\": dial tcp: i/o timeout"
2023-09-05T09:30:29.657632797Z stderr F time="2023-09-05T09:30:29.657Z" level=error msg="executor error: artifact mlops-core-2-output_dir failed to load: failed to get file: Get \"http://minio-service.kubeflow.svc.cluster.local:9000/mlpipeline/?location=\": dial tcp: i/o timeout"
```

### Version

v3.3.8

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
Just need to bring down minio for couple of minutes to simulate network outage for few seconds.
```

### Logs from the workflow controller

```text
NA
```

### Logs from in your workflow's wait container

```text
2023-09-05T10:30:46+01:00	2023-09-05T09:30:29.657653647Z stderr F time="2023-09-05T09:30:29.657Z" level=fatal msg="artifact mlops-core-2-output_dir failed to load: failed to get file: Get \"http://minio-service.kubeflow.svc.cluster.local:9000/mlpipeline/?location=\": dial tcp: i/o timeout"
2023-09-05T10:30:46+01:00	2023-09-05T09:30:29.657645967Z stderr F time="2023-09-05T09:30:29.657Z" level=info msg="Alloc=6243 TotalAlloc=11416 Sys=23762 NumGC=4 Goroutines=5"
2023-09-05T10:30:46+01:00	2023-09-05T09:30:29.657632797Z stderr F time="2023-09-05T09:30:29.657Z" level=error msg="executor error: artifact mlops-core-2-output_dir failed to load: failed to get file: Get \"http://minio-service.kubeflow.svc.cluster.local:9000/mlpipeline/?location=\": dial tcp: i/o timeout"
2023-09-05T10:30:29+01:00	2023-09-05T09:30:29.657599935Z stderr F time="2023-09-05T09:30:29.657Z" level=warning msg="Non-transient error: Get \"http://minio-service.kubeflow.svc.cluster.local:9000/mlpipeline/?location=\": dial tcp: i/o timeout"
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.656401542Z stderr F time="2023-09-05T09:29:59.656Z" level=info msg="Getting file from s3" bucket=mlpipeline endpoint="minio-service.kubeflow.svc.cluster.local:9000" key=artifacts/abc-1-74521065/2023/09/05/abc4wf4g-1-74521065-2101737519/mlops-core-2-output_dir.tgz path=/argo/inputs/artifacts/mlops-core-2-output_dir.tmp
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.656394078Z stderr F time="2023-09-05T09:29:59.656Z" level=info msg="Creating minio client using static credentials" endpoint="minio-service.kubeflow.svc.cluster.local:9000"
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.656391144Z stderr F time="2023-09-05T09:29:59.656Z" level=info msg="S3 Load path: /argo/inputs/artifacts/mlops-core-2-output_dir.tmp, key: artifacts/abc4wf4g-1-74521065/2023/09/05/rewardtrainingnightly4wf4g-1-74521065-2101737519/mlops-core-2-output_dir.tgz"
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.656388158Z stderr F time="2023-09-05T09:29:59.656Z" level=info msg="Downloading artifact: mlops-core-2-output_dir"
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.656369273Z stderr F time="2023-09-05T09:29:59.656Z" level=info msg="Start loading input artifacts..."
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.605900927Z stderr F time="2023-09-05T09:29:59.605Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.605859557Z stderr F time="2023-09-05T09:29:59.605Z" level=info msg="Creating a emissary executor"
2023-09-05T10:30:04+01:00	2023-09-05T09:29:59.602802461Z stderr F time="2023-09-05T09:29:59.602Z" level=info msg="Starting Workflow Executor" executorType=emissary version=v3.3.8
```

**Comment by terrytangyuan** [comment]
Can you try the newer versions? I tested that `Get \"http://minio-service.kubeflow.svc.cluster.local:9000/mlpipeline/?location=\": dial tcp: i/o timeout"` should be a transient error already.

**Comment by Sharathmk99** [comment]
Hi @terrytangyuan , I’m using Argo part of Kubeflow. Not sure if it’s easy to upgrade the version, let me see if it works if I just upgrade. Thanks.

**Comment by agilgur5** [comment]
> v3.3.8

3.3.8 should already be handling a [request timeout as a transient error](https://github.com/argoproj/argo-workflows/blob/v3.3.8/workflow/artifacts/s3/errors.go#L13).

Perhaps this is a different kind of timeout? 🤔

**Comment by agilgur5** [comment]
> Perhaps this is a different kind of timeout? 🤔

Upstream MinIO only lists [request timeout](https://github.com/minio/minio-go/blob/a750ce3dc18eec7e4bfb484872527920256b46ef/retry.go#L93), so this might be an upstream issue if it's not considering `dial tcp: i/o timeout` as an transient error.

**Comment by Sharathmk99** [comment]
Looks like it’s upstream responding timeout. Anyway to handle it as
transient error?
Thanks.

**Comment by terrytangyuan** [comment]
It should already be covered https://github.com/argoproj/argo-workflows/blob/master/util/errors/errors.go#L74

**Comment by Sharathmk99** [comment]
That’s interesting. How come still I get non-transient error? Any leads to
debug?

**Comment by terrytangyuan** [comment]
I would suggest trying out the newer versions to see if it's fixed.

**Comment by agilgur5** [comment]
I think this should have been fixed by #12191

**Comment by terrytangyuan** [comment]
Closing for now. Feel free to re-open if you see this still happening in new versions.
